app/src/main/kotlin/compiler/parser/analysis/GrammarAnalysis.kt

```diff
@@ -184,4 +165,20 @@ class GrammarAnalysis<S : Comparable<S>> {
 
         return result
     }
+
+    fun computeFirstPlus(nullable: Set<S>, first: Map<S, Set<S>>, follow: Map<S, Set<S>>): Map<S, Set<S>> {
+        val firstPlus: MutableMap<S, MutableSet<S>> = HashMap()
+
+        for ((symbol, firstOfSymbol) in first) {
+            firstPlus.putIfAbsent(symbol, HashSet())
+            firstPlus[symbol]!!.addAll(firstOfSymbol)
+        }
+
+        for (symbol in nullable) {
+            firstPlus.putIfAbsent(symbol, HashSet())
+            firstPlus[symbol]!!.addAll(follow[symbol]!!)
+        }
+
+        return firstPlus
```

Comment:
I feel like this function could be shorter and more readable with something like `return grammar.getSymbols().associateWith` and `union` to closer match the mathematical definition.